core-bundle/contao/drivers/DC_Table.php

```diff
@@ -2209,62 +2195,52 @@ public function edit($intId=null, $ajaxId=null)
-				$this->strTable,
-				(bool) $this->ptable,
-				$security->isGranted(ContaoCorePermissions::DC_PREFIX . $this->strTable, new CreateAction($this->strTable, $this->addDynamicPtable(array('pid' => $this->intCurrentPid)))),
-				$security->isGranted(ContaoCorePermissions::DC_PREFIX . $this->strTable, new CreateAction($this->strTable, array_replace($currentRecord, array('id' => null, 'sorting' => null)))),
-				$this
-			);
-
-		// Add the buttons and end the form
-		$return .= '
-</div>
-  ' . $strButtons . '
-</form>';
-
-		$strVersionField = '';
-
-		// Store the current version number (see #8412)
-		if ($intLatestVersion !== null)
-		{
-			$strVersionField = '
-<input type="hidden" name="VERSION_NUMBER" value="' . $intLatestVersion . '">';
-		}
-
+		$parameters['form'] = array(
+			'multipart' => $this->blnUploadable,
+			'onsubmit' => $this->onsubmit,
+			'buttons' => System::getContainer()
+				->get('contao.data_container.buttons_builder')
+				->generateEditButtons(
+					$this->strTable,
+					(bool) $this->ptable,
+					$security->isGranted(ContaoCorePermissions::DC_PREFIX . $this->strTable, new CreateAction($this->strTable, $this->addDynamicPtable(array('pid' => $this->intCurrentPid)))),
+					$security->isGranted(ContaoCorePermissions::DC_PREFIX . $this->strTable, new CreateAction($this->strTable, array_replace($currentRecord, array('id' => null, 'sorting' => null)))),
+					$this
+				),
+		);
+
+		// Back button
 		$strBackUrl = $this->getReferer(true);
 
 		if ((string) $currentRecord['tstamp'] === '0')
 		{
 			$strBackUrl = preg_replace('/&(?:amp;)?revise=[^&]+|$/', '&amp;revise=' . $this->strTable . '.' . ((int) $this->intId), $strBackUrl, 1);
 		}
 
-		// Begin the form (-> DO NOT CHANGE THIS ORDER -> this way the onsubmit attribute of the form can be changed by a field)
-		$return = $version . Message::generate() . ($this->noReload ? '
-<p class="tl_error">' . $GLOBALS['TL_LANG']['ERR']['submit'] . '</p>' : '') . (Input::get('nb') ? '' : System::getContainer()->get('contao.data_container.global_operations_builder')->initialize($this->strTable)->addBackButton($strBackUrl)) . '
-<form id="' . $this->strTable . '" class="tl_form tl_edit_form" method="post" enctype="' . ($this->blnUploadable ? 'multipart/form-data' : 'application/x-www-form-urlencoded') . '"' . (!empty($this->onsubmit) ? ' onsubmit="' . implode(' ', $this->onsubmit) . '"' : '') . '>
-<div class="tl_formbody_edit">
-<input type="hidden" name="FORM_SUBMIT" value="' . $this->strTable . '">
-<input type="hidden" name="REQUEST_TOKEN" value="' . htmlspecialchars(System::getContainer()->get('contao.csrf.token_manager')->getDefaultTokenValue(), ENT_QUOTES | ENT_SUBSTITUTE | ENT_HTML5) . '">' . $strVersionField . $return;
+		$parameters['back_button'] = (Input::get('nb') ?
+			'' :
+			System::getContainer()
+				->get('contao.data_container.global_operations_builder')
+				->initialize($this->strTable)
+				->addBackButton($strBackUrl))
+				->__toString()
```

Comment:
```suggestion
			System::getContainer()
				->get('contao.data_container.global_operations_builder')
				->initialize($this->strTable)
				->addBackButton($strBackUrl))
```

<img width="897" height="592" alt="Image" src="https://github.com/user-attachments/assets/08d277bb-e588-4ec5-a882-a8b6dd66e37a" />

Got this exception when clicking on the _edit archive_ popup operation inside the news list view (inside the archive). Not sure why it is happening, but we do not call `__toString` when generating the edit operations, and removing it seems to fix the issue.

Reply:
Thanks for testing! There were actually wrong extra parentheses (see the line above the error) :see_no_evil: Hard to see. 917bcd421f8d04739116f15cc246362063463ff6 fixes it.

I think we should keep the `__toString()` call for two reasons: this will make sure the buttons get generated at this exact point and the template parameter will always be of the same type (string). 

If you prefer, we could alternatively make it `DataContainerGlobalOperationsBuilder|null` and test for it in the template.

Reply:
Friendly ping @aschempp - I'd like to get this one going soon, so that the work on the next methods can continue.